### PR TITLE
Add native method stubs of jdk.jfr.internal.JVM on different versions

### PR DESCRIPTION
**Pull request by thallium**

### CODE REVIEW

**Comment by thallium** [comment]
@tajila FYI

**Comment by tajila** [comment]
Can you do it in a way such that the JDK11 version is includes everything in JDK11 and onwards. That way JDK17 is just the new methods in JDK17 and onwards, same for JDK21 and 23

**Comment by thallium** [comment]
The problem is some methods also got removed in a later version...

**Comment by tajila** [comment]
> The problem is some methods also got removed in a later version...

Is there a common set of methods that we can put in the base version of the file? I'd like to avoid code duplication.

**Comment by thallium** [comment]
> Is there a common set of methods that we can put in the base version of the file?

Yes, but one thing to note is that on JDK23 every method is changed to static, although I think function with old signature can still be linked.

**Comment by tajila** [comment]
> Yes, but one thing to note is that on JDK23 every method is changed to static, although I think function with old signature can still be linked.

you can define the first params like

```
#ifdef JDK23
#define JFR_NATIVES_PARAM JNIEnv *env, jclass clazz
#else
#define JFR_NATIVES_PARAM JNIEnv *env, jobject obj
#endif
```

File: runtime/jcl/common/jdk_jfr_internal_JVM_jdk23.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+void JNICALL
+Java_jdk_jfr_internal_JVM_flush__Ljdk_jfr_internal_event_EventWriter_2II(JNIEnv *env, jclass cls, jobject writer, jint uncommittedSize, jint requestedSize)
+{
+	// TODO: implementation
+}
+
+void JNICALL
+Java_jdk_jfr_internal_JVM_flush__(JNIEnv *env, jclass cls)
```
Comment (tajila) [inline runtime/jcl/common/jdk_jfr_internal_JVM_jdk23.cpp:48]: This one appears in multiple places. You could put it in common like this
```
#ifdef 11
Java_jdk_jfr_internal_JVM_flush
#else
Java_jdk_jfr_internal_JVM_flush__
#endif
```

File: runtime/jcl/common/jdk_jfr_internal_JVM_jdk11.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+void JNICALL
+Java_jdk_jfr_internal_JVM_setSampleThreads(JNIEnv *env, jobject obj, jboolean sampleThreads)
+{
+	// TODO: implementation
+}
+
+jlong JNICALL
+Java_jdk_jfr_internal_JVM_getTypeId(JNIEnv *env, jobject obj, jclass clazz)
```
Comment (tajila) [inline runtime/jcl/common/jdk_jfr_internal_JVM_jdk11.cpp:47]: In cases where the only difference is static vs instance you can do something like this:

```
#ifdef JDK23
#define JFR_NATIVES_PARAM JNIEnv *env, jclass unused
#else
#define JFR_NATIVES_PARAM JNIEnv *env, jobject unused
#endif
```

File: runtime/jcl/common/jdk_jfr_internal_JVM_jdk17.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+jobject JNICALL
+Java_jdk_jfr_internal_JVM_getHandler(JNIEnv *env, jobject obj, jclass eventClass)
+{
+	// TODO: implementation
+	return NULL;
+}
+
+}
```
Comment (keithc-ca) [inline runtime/jcl/common/jdk_jfr_internal_JVM_jdk17.cpp:47]: Please add a comment to each brace that closes `extern "C"`:
```c
} /* extern "C" */
```

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk11

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk17

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Comment by tajila** [comment]
jenkins test sanity xlinux jdk11

File: runtime/jcl/exports.cmake [hunk 4]
```diff
@@ -695,73 +695,261 @@ if(J9VM_OPT_JFR)
...
+		omr_add_exports(jclse
+			Java_jdk_jfr_internal_JVM_registerNatives
+			Java_jdk_jfr_internal_JVM_beginRecording
+			Java_jdk_jfr_internal_JVM_counterTime
+			Java_jdk_jfr_internal_JVM_emitEvent
+			Java_jdk_jfr_internal_JVM_endRecording
+			Java_jdk_jfr_internal_JVM_getAllEventClasses
+			Java_jdk_jfr_internal_JVM_getUnloadedEventClassCount
+			Java_jdk_jfr_internal_JVM_getClassId
+			Java_jdk_jfr_internal_JVM_getClassIdNonIntrinsic
```
Comment (keithc-ca) [inline runtime/jcl/exports.cmake:709]: Please sort each list of exports alphabetically.

File: runtime/jcl/module.xml [hunk 5]
```diff
@@ -142,8 +142,21 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
 			<group name="se20">
 				<include-if condition="spec.java20"/>
 			</group>
-			<group name="jfr">
+			<group name="jfr_jdk11">
+				<include-if condition="spec.flags.opt_jfr"/>
+				<include-if condition="spec.java11"/>
+			</group>
```
Comment (keithc-ca) [inline runtime/jcl/module.xml:148]: This is not consistent with the cmake version (`spec.java11` means `JAVA_SPEC_VERSION >= 11`); an exclude is needed:
```
				<exclude-if condition="spec.java12"/>
```
Similarly for the jdk17 and jdk21 groups.

Please also remove the unwanted space before `>` in `runtime/jcl/uma/jfr_objects.xml`:
```
	<object name="jdk_jfr_internal_JVM">
```

File: runtime/jcl/uma/jfr_jdk11_exports.xml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+-->
+<exports group="jfr_jdk11">
+	<export name="Java_jdk_jfr_internal_JVM_registerNatives" />
+	<export name="Java_jdk_jfr_internal_JVM_beginRecording" />
+	<export name="Java_jdk_jfr_internal_JVM_counterTime" />
+	<export name="Java_jdk_jfr_internal_JVM_emitEvent" />
+	<export name="Java_jdk_jfr_internal_JVM_endRecording" />
+	<export name="Java_jdk_jfr_internal_JVM_getAllEventClasses" />
+	<export name="Java_jdk_jfr_internal_JVM_getUnloadedEventClassCount" />
```
Comment (keithc-ca) [inline runtime/jcl/uma/jfr_jdk11_exports.xml:29]: Please sort export lists alphabetically.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/jcl/module.xml (hunk 5 above)
Comment (thallium) [reply runtime/jcl/module.xml:148]: For JDK23, should I keep it as is?

File: runtime/jcl/common/jdk_jfr_internal_JVM_common.cpp [hunk 7]
```diff
@@ -38,94 +38,87 @@ extern "C" {
 #define JFR_STRING_BUFFER 256
 
 void JNICALL
-Java_jdk_jfr_internal_JVM_registerNatives(JNIEnv *env, jclass clazz)
+Java_jdk_jfr_internal_JVM_registerNatives(JNIEnv *env, jclass cls)
```
[the remark this answers is not in the document]
Comment (thallium) [reply runtime/jcl/common/jdk_jfr_internal_JVM_common.cpp]: For functions which already have a parameter called clazz, I'm just keeping it as cls for now.

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk17

File: runtime/jcl/common/jdk_jfr_internal_JVM_common.cpp [hunk 8]
```diff
@@ -44,88 +44,81 @@ Java_jdk_jfr_internal_JVM_registerNatives(JNIEnv *env, jclass clazz)
...
-Java_jdk_jfr_internal_JVM_getUnloadedEventClassCount(JNIEnv *env, jclass clazz)
+Java_jdk_jfr_internal_JVM_getUnloadedEventClassCount(JNIEnv *env, jobject obj)
 {
 	// TODO: implementation
 	return 0;
 }
 
 jlong JNICALL
-Java_jdk_jfr_internal_JVM_getClassId(JNIEnv *env, jclass clz, jobject clazz)
+Java_jdk_jfr_internal_JVM_getClassId(JNIEnv *env, jclass cls, jclass clazz)
```
Comment (keithc-ca) [inline runtime/jcl/common/jdk_jfr_internal_JVM_common.cpp:93]: I think `clazz` should be used consistently for the context class; the second `jclass` parameter should get a new name, like "subjectClass" or "targetClass" that clarifies its purpose.

File: runtime/jcl/common/jdk_jfr_internal_JVM_jdk11.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "jni.h"
+
+extern "C" {
+
+jlong JNICALL
+Java_jdk_jfr_internal_JVM_getClassIdNonIntrinsic(JNIEnv *env, jclass cls, jclass clazz)
```
Comment (keithc-ca) [inline runtime/jcl/common/jdk_jfr_internal_JVM_jdk11.cpp:28]: Again. use `clazz` for the context, and chose a suitable name for the explicit parameter.

File: runtime/jcl/exports.cmake [hunk 10]
```diff
@@ -695,73 +695,261 @@ if(J9VM_OPT_JFR)
...
+			Java_jdk_jfr_internal_JVM_setStackDepth
+			Java_jdk_jfr_internal_JVM_setStackTraceEnabled
+			Java_jdk_jfr_internal_JVM_setThreadBufferSize
+			Java_jdk_jfr_internal_JVM_setThreshold
+			Java_jdk_jfr_internal_JVM_setThrottle
+			Java_jdk_jfr_internal_JVM_shouldRotateDisk
+			Java_jdk_jfr_internal_JVM_storeMetadataDescriptor
+			Java_jdk_jfr_internal_JVM_subscribeLogLevel
+			Java_jdk_jfr_internal_JVM_uncaughtException
+			java_jdk_jfr_internal_JVM_emitDataLoss
```
Comment (keithc-ca) [inline runtime/jcl/exports.cmake:877]: Please keep the list of exports sorted; this belongs before line 820 (`Java_jdk_jfr_internal_JVM_emitEvent`).

File: runtime/jcl/module.xml [hunk 11]
```diff
@@ -142,8 +142,21 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-ex
 			<group name="se20">
 				<include-if condition="spec.java20"/>
 			</group>
-			<group name="jfr">
+			<group name="jfr_jdk11">
+				<include-if condition="spec.flags.opt_jfr"/>
+				<exclude-if condition="spec.java12"/>
```
Comment (keithc-ca) [inline runtime/jcl/module.xml:147]: This group also needs `<include-if condition="spec.java11"/>` to be consistent with cmake builds.
Similar comment applies to `jfr_jdk17` and `jfr_jdk21`.

**Review by keithc-ca — CHANGES_REQUESTED** [review]
Please also fix `jfr_objects.xml` by removing the space before `>` on line 23.

**Comment by tajila** [comment]
@keithc-ca Please take a look at these changes

**Comment by thallium** [comment]
@keithc-ca Addressed your feedbacks

File: runtime/jcl/common/jdk_jfr_internal_JVM_jdk21andUp.cpp [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+jlong JNICALL
+Java_jdk_jfr_internal_JVM_hostTotalMemory(JNIEnv *env, jobject obj)
+{
+	// TODO: implementation
+	return 0;
+}
+
+}
+ /* extern "C" */
```
Comment (keithc-ca) [inline runtime/jcl/common/jdk_jfr_internal_JVM_jdk21andUp.cpp:91]: Please join these lines (as it the custom elsewhere).
```c
} /* extern "C" */
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/jcl/common/jdk_jfr_internal_JVM_jdk21andUp.cpp (hunk 12 above)
Comment (thallium) [reply runtime/jcl/common/jdk_jfr_internal_JVM_jdk21andUp.cpp]: Sorry, fixed

**Comment by keithc-ca** [comment]
Jenkins test sanity.functional zlinux jdk8,jdk11,jdk17,jdk21,jdk23

**Review by keithc-ca — APPROVED** [review]
The test failures are with CRIU or JIT - they appear unrelaetd to this change.

**Comment by tajila** [comment]
@thallium Please backport to 0.49

**Pull request merged**